### ISSUE TITLE
Feature request: syntax introspection that includes line numbers

### ISSUE THREAD

**Issue by cpitclaudel** [coq/coq#11181]
#### Description of the problem

I'm writing an EDSL in Coq.  Users write untyped programs, which are then typechecked.
At the moment, when a program contains a type error, I create a new version of the AST containing a special error node, and display that to the user.  Then they can see where the error was from by looking for `>>> error-here <<<` in the output.

This is not satisfactory for three reasons: first, when terms are large, it becomes hard to locate the error.  Second, not all of my notations are reprintable, so looking at the term with the error isn't always pleasant.  Third, my language relies on macros (Coq functions that expand into syntax), and so typechecked terms are often rather large and not very readable

Additionally, after typechecking and compilation, I produce low-level code that I'd like to annotate with positions to aid debugging.  My AST-annotation trick doesn't work for that purpose.

Hence, I'd love to have a way to get a copy of a term with syntax annotations, such as positions.  The simplest way would be to introduce magic __FILE__, __LINE__, and __COLUMN__ constants that would be replaced with the current file name, line, and column number at parsing time (and, when used in a notation, when the notation is applied).

An alternative would be to add a feature in Ltac2 to get a copy of the AST with position annotations.

Do any of these sound realistic? I can provide more examples of use cases if needed.

Thanks!

**Comment by Zimmi48** [comment]
Could this also be the kind of things that could be granted via things like SerAPI? cc @ejgallego

**Comment by cpitclaudel** [comment]
No, I think not.  I need access to the line numbers from Gallina and Ltac programs, not from an outside UI.

**Comment by ejgallego** [comment]
Indeed; I wonder what metaCoq can do here.

**Comment by ppedrot** [comment]
Untyped terms in the ML implementation contain locations, it's not very hard to expose the corresponding Ltac2 API.

**Comment by cpitclaudel** [comment]
:heart_eyes:

**Comment by cpitclaudel** [comment]
@ppedrot, any updates on this?

**Comment by cpitclaudel** [comment]
Wonderful!  That's super exciting.  I left an example on the other thread.

**Comment by SkySkimmer** [comment]
Actually I didn't read the feature request very carefully, I'm not sure ltac2 is the right place for this, it may be nicer as some kind of Notation feature.
IDK

**Comment by cpitclaudel** [comment]
I like the idea of Ltac2, FWIW; dealing with locations and syntax seems within scope for a metaprogramming language?
